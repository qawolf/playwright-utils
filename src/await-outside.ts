declare module 'await-outside' {
  import { REPLServer } from 'repl';

  export function addAwaitOutsideToReplServer(repl: REPLServer): void;
}
