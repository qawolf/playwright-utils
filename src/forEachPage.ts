import { BrowserContext, Page } from 'playwright-core';

export const forEachPage = async (
  context: BrowserContext,
  pageFn: (page: Page) => void | Promise<void>,
): Promise<void> => {
  // XXX remove cast after playwright fixes types
  (context as any).on('page', async event => pageFn(await event.page()));
  const pages = await context.pages();
  await Promise.all([pages.map(page => pageFn(page))]);
};
