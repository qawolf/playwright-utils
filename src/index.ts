// context utils
export { forEachPage } from './context/forEachPage';
export { saveArtifacts } from './context/saveArtifacts';

// page utils
export { initEvaluateScript } from './page/initEvaluateScript';
export { interceptConsoleLogs } from './page/interceptConsoleLogs';
export { saveConsoleLogs } from './page/saveConsoleLogs';
export { saveState } from './page/saveState';
export { scroll } from './page/scroll';
export { setState } from './page/setState';

// global utils
export { getLaunchOptions, launch } from './launch';
export { repl, ReplContext } from './repl';

// for internal use
export { waitFor } from './waitFor';
export { indexPages, IndexedPage } from './context/indexPages';
export { waitForPage } from './context/waitForPage';
