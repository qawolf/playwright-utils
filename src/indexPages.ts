import Debug from 'debug';
import { BrowserContext, Page } from 'playwright-core';

const debug = Debug('playwright-utils:indexPages');

type IndexedBrowserContext = BrowserContext & {
  _putils_indexed: boolean;
};

export type IndexedPage = Page & {
  createdIndex: number;
};

export const indexPages = async (context: BrowserContext) => {
  /**
   * Set page.createdIndex on pages.
   */
  if ((context as IndexedBrowserContext)._putils_indexed) return;
  (context as IndexedBrowserContext)._putils_indexed = true;

  let index = 0;

  const pages = await context.pages();
  if (pages.length > 1) {
    throw new Error('Cannot index pages when >1 exists');
  }

  if (pages[0]) {
    debug(`index existing page ${index}`);
    (pages[0] as IndexedPage).createdIndex = index++;
  }

  // TODO remove cast after PR is merged
  (context as any).on('page', async event => {
    debug(`index created page ${index}`);
    const page: IndexedPage = await event.page();
    page.createdIndex = index++;
  });
};
