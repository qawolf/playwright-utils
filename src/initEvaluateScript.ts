import { Page } from 'playwright-core';

export const initEvaluateScript = (page: Page, script: string | Function) =>
  // TODO fix cast
  Promise.all([page.addInitScript(script), page.evaluate(script as any)]);
