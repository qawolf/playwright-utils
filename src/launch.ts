import * as Debug from 'debug';
import { platform } from 'os';
import playwright, { Browser } from 'playwright-core';
import { LaunchOptions as PlaywrightLaunchOptions } from 'playwright-core/lib/server/browserType';
import { isNullOrUndefined } from 'util';

const debug = Debug('playwright-utils:launch');

type BrowserName = 'chromium' | 'firefox' | 'webkit';

export type LaunchOptions = PlaywrightLaunchOptions & {
  browser?: BrowserName;
};

const parseBool = (value: string | undefined) => {
  const lowerCaseValue = (value || '').toLowerCase();
  return ['1', 't', 'true'].includes(lowerCaseValue);
};

const parseBrowserName = (name?: string): string => {
  if (name === 'firefox' || name === 'webkit') return name;
  return 'chromium';
};

export const getLaunchOptions = (options: LaunchOptions = {}) => {
  const launchOptions = { ...options };

  const headlessEnv = process.env.QAW_HEADLESS;
  if (isNullOrUndefined(options.headless) && !isNullOrUndefined(headlessEnv)) {
    launchOptions.headless = parseBool(headlessEnv);
  }

  const browserName = parseBrowserName(process.env.QAW_BROWSER);

  if (isNullOrUndefined(options.args)) {
    let args: string[] = [];
    if (browserName === 'chromium') {
      args = [
        '--disable-dev-shm-usage',
        '--no-default-browser-check',
        '--window-position=0,0',
      ];

      if (platform() === 'linux') {
        // We use --no-sandbox because we cannot change the USER for certain CIs (like GitHub).
        // "Ensure your Dockerfile does not set the USER instruction, otherwise you will not be able to access GITHUB_WORKSPACE"
        args.push('--no-sandbox');
      }
    }
  }

  return { ...launchOptions, browserName };
};

export const launch = async (options: LaunchOptions = {}): Promise<Browser> => {
  const launchOptions = getLaunchOptions(options);
  debug('launch %j', launchOptions);

  const browser = await playwright[launchOptions.browserName].launch(
    launchOptions,
  );
  return browser;
};
