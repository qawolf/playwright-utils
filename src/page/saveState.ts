import { ensureFile, writeJSON } from 'fs-extra';
import { Page } from 'playwright';
import { NetworkCookie } from 'playwright-core/lib/network';

export interface State {
  cookies: NetworkCookie[];
  localStorage: object;
  sessionStorage: object;
}

const getState = async (page: Page): Promise<State> => {
  const context = page.context();
  const cookies = await context.cookies();
  const { localStorage, sessionStorage } = await page.evaluate(() => {
    return {
      localStorage: { ...localStorage },
      sessionStorage: { ...sessionStorage },
    };
  });

  return { cookies, localStorage, sessionStorage };
};

export const saveState = async (
  page: Page,
  savePath: string,
): Promise<void> => {
  const state = await getState(page);

  await ensureFile(savePath);
  return writeJSON(savePath, state);
};
