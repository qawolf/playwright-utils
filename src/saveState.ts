import { ensureFile, writeJSON } from 'fs-extra';
import { BrowserContext, Page } from 'playwright-core';
import { NetworkCookie } from 'playwright-core/lib/network';

interface GetStateOptions {
  context: BrowserContext;
  page: Page;
}

export interface State {
  cookies: NetworkCookie[];
  localStorage: object;
  sessionStorage: object;
}

interface SaveStateOptions extends GetStateOptions {
  savePath: string;
}

const getState = async ({ context, page }: GetStateOptions): Promise<State> => {
  const cookies = await context.cookies();
  const { localStorage, sessionStorage } = await page.evaluate(() => {
    return { localStorage, sessionStorage };
  });

  return { cookies, localStorage, sessionStorage };
};

export const saveState = async ({
  context,
  page,
  savePath,
}: SaveStateOptions): Promise<void> => {
  const state = await getState({ context, page });

  await ensureFile(savePath);
  return writeJSON(savePath, state);
};
