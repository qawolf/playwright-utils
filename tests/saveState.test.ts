import { pathExists, readJSON } from 'fs-extra';
import { tmpdir } from 'os';
import { join } from 'path';
import { Browser } from 'playwright';
import { launch } from '../src/launch';
import { saveState } from '../src/saveState';
import { randomString } from './utils';

const COOKIE = {
  sameSite: 'Lax' as 'Lax',
  name: 'csrf_same_site',
  value: '1',
  domain: '.twitter.com',
  path: '/',
  expires: 1615052185.041635,
  httpOnly: true,
  secure: true,
};

describe('saveState', () => {
  let browser: Browser;

  beforeAll(async () => {
    browser = await launch();
  });

  afterAll(() => browser.close());

  it('saves state to the specified file', async () => {
    const savePath = join(tmpdir(), randomString(), 'state.json');

    const page = await browser.newPage();
    await page.goto('http://localhost:5000');

    await page.context().setCookies([COOKIE]);
    await page.evaluate(() => {
      localStorage.setItem('hello', 'world');
      sessionStorage.setItem('in', 'sessionStorage');
    });

    await saveState(page, savePath);

    const isStateSaved = await pathExists(savePath);
    expect(isStateSaved).toBe(true);

    const state = await readJSON(savePath);
    expect(state).toMatchSnapshot();
  });
});
