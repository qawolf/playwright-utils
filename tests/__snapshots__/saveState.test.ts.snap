// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`saveState saves state to specified file 1`] = `
Object {
  "cookies": Array [
    Object {
      "domain": ".twitter.com",
      "expires": 1615052185.041635,
      "httpOnly": true,
      "name": "csrf_same_site",
      "path": "/",
      "sameSite": "Lax",
      "secure": true,
      "session": false,
      "value": "1",
    },
  ],
  "localStorage": Object {
    "hello": "world",
  },
  "sessionStorage": Object {
    "in": "sessionStorage",
  },
}
`;
